pallets/erc20-xcm-bridge/src/xcm_holding_ext.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2019-2022 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Module that provides types to extend xcm holding.
+
+use crate::erc20_matcher::Erc20Matcher;
+use core::marker::PhantomData;
+use sp_core::{H160, U256};
+use sp_std::collections::btree_map::BTreeMap;
+use sp_std::vec::Vec;
+use xcm_executor::traits::{DropAssets, MatchesFungibles};
+
+environmental::environmental!(XCM_HOLDING_ERC20_ORIGINS: XcmHoldingErc20sOrigins);
+
+#[cfg_attr(test, derive(PartialEq, Debug))]
+pub(crate) enum DrainError {
+	NotEnoughFounds,
+	AssetNotFound,
+}
+
+/// Xcm holding erc20 origins extension.
+/// This extension track down the origin of alls erc20 tokens in the xcm holding.
+#[derive(Default)]
+pub(crate) struct XcmHoldingErc20sOrigins {
+	map: BTreeMap<H160, Vec<(H160, U256)>>,
```

Comment:
maybe not strictly necessary but should we consider `BoundedVec` instead?

Reply:
The implementation of xcm holding ans polkadot doesn't give a limit to the number of assets, so to be compatible you shouldn't give a limit here.

It is rather the role of the XCM Barrier to limit the number of different assets and the number of instructions.